### ISSUE TITLE
Error mapping node with custom CYPHER RETURN leaving out relations in SDN 7.1.1

### ISSUE THREAD

**Issue by ruempel** [spring-projects/spring-data-neo4j#2749]
# Setup

I use an application with Spring Boot `3.1.0` (which brings Spring Data Neo4j `7.1.0`, but the same behavior occurs when using SDN `7.1.1`) with a Neo4j database `5.8.0`. Further, I have a node like this:

```java
@Node
@Data
public class SourceNode {
    @Id
    @GeneratedValue(generatorClass = UUIDStringGenerator.class)
    private String uuid;

    @Relationship(type = "RELATION_NAME", direction = Relationship.Direction.OUTGOING)
    private List<TargetNode> randomRelationName = new ArrayList<>();
}
```

and a repository with custom CYPHER queries like this:

```java
public interface SourceNodeRepository extends CrudRepository<SourceNode, String> {
    @Query("MATCH (s:SourceNode) RETURN s{.uuid}")
    List<SourceNode> getAllWithoutRelations();

    @Query("MATCH (s:SourceNode) RETURN s{.uuid, __elementId__: toString(id(s))}")
    List<SourceNode> getAllWithoutRelationsFixed();
}
```

# Suspected Bug

When I execute `getAllWithoutRelations()` to fetch the source nodes (without intentionally mapping the relations, since I want to continue with very lean objects), I get:

```
org.springframework.data.mapping.MappingException: Error mapping Record<{role: {uuid: "xxxxx"}}>
...
Caused by: java.util.NoSuchElementException: No value present
```

Is it a bug in Spring Data Neo4j `7.1.0` (resp. `7.1.1`) or how can I prevent this problem properly?

When using Spring Boot `3.0.6` bringing Spring Data Neo4j `7.0.5` the query execution works without problems.

# Ugly Workaround

I found out, that the problem relates to the source node's internal Neo4j ID, which cannot be determined based on the actual query when executing the custom query. By adding `__elementId__: toString(id(s))` (see repository method `getAllWithoutRelationsFixed()`) to the return object, the problem can be solved, but I doubt that this is a good solution.

**Comment by meistermeier** [comment]
Did this work before for you? I am asking because this _should_ never have worked (or was at least a supported feature). I am even wondering that the workaround fixes this for you. Spring Data Neo4j requires usually a complete node or a map projection including _labels_, and _elementId_ or _id_ (https://docs.spring.io/spring-data/neo4j/docs/current/reference/html/#query-creation.load on the following section on custom queries).
Basically the fix in your custom query is not yet complete because it is missing out the _labels_.
Also, you could just use `elementId(s)` and not do the conversion from (legacy) id to string, if you are using a current version of Neo4j (5.x).

I am currently seeing that we are missing out the _elementId_ in the documentation but this is something I will add right now for the next release.

**Comment by ruempel** [comment]
> Did this work before for you? I am asking because this _should_ never have worked (or was at least a supported feature). I am even wondering that the workaround fixes this for you. Spring Data Neo4j requires usually a complete node or a map projection including _labels_, and _elementId_ or _id_ (https://docs.spring.io/spring-data/neo4j/docs/current/reference/html/#query-creation.load on the following section on custom queries).

Yes, the query used in `getAllWithoutRelations()` works until SDN 7.0.7 and does not work with SDN 7.1.0 and 7.1.1.

> Basically the fix in your custom query is not yet complete because it is missing out the _labels_. Also, you could just use `elementId(s)` and not do the conversion from (legacy) id to string, if you are using a current version of Neo4j (5.x).

Thank you for the hint. I just tried it out and `elementId(s)` also works instead of `toString(id(s))`, but this whole kind of fix is not what we actually want to use.

We require these kinds of queries to cut off unwanted fields and relations (to selectively prevent nested objects and prevent loops) and use the same mapped object model at the same time.

**Comment by meistermeier** [comment]
A simple 
```
@Node
public class Thing {
  @Id @GeneratedValue(UUIDStringGenerator.class) private String theId;
}
```
with
```
interface ThingRepository extends Neo4jRepository<Thing, String> {
	@Query("MATCH (n) return n{.theId}")
	List<Thing> getThings();
}
```
works fine for me in 7.0 *and* 7.1.
Could you provide the complete stack trace?

**Comment by ruempel** [comment]
> A simple
> 
> ```
> @Node
> public class Thing {
>   @Id @GeneratedValue(UUIDStringGenerator.class) private String theId;
> }
> ```
> 
> with
> 
> ```
> interface ThingRepository extends Neo4jRepository<Thing, String> {
> 	@Query("MATCH (n) return n{.theId}")
> 	List<Thing> getThings();
> }
> ```
> 
> works fine for me in 7.0 _and_ 7.1. Could you provide the complete stack trace?

The relation makes the difference, see `SourceNode.randomRelationName` in my original example.

**Comment by meistermeier** [comment]
There is now a 7.1.2-GH-2749-SNAPSHOT available. 
Would be good to get feedback if this solves your problem.

**Comment by ruempel** [comment]
Yes, it solves the described problem. Thank you for reacting quickly.
